jackson/src/main/java/feign/jackson/JacksonIteratorDecoder.java

```diff
@@ -0,0 +1,168 @@
+ * <pre><code>
+ * Feign.builder()
+ *   .decoder(JacksonIteratorDecoder.Factory.create())
+ *   .doNotCloseAfterDecode() // Required to fetch the iterator after the response is processed, need to be close
+ *   .target(GitHub.class, "https://api.github.com");
+ * interface GitHub {
+ *  {@literal @}RequestLine("GET /repos/{owner}/{repo}/contributors")
+ *   Iterator<Contributor> contributors(@Param("owner") String owner, @Param("repo") String repo);
+ * }</code></pre>
+ */
+public class JacksonIteratorDecoder implements Decoder {
+
+  private final ObjectMapper mapper;
+
+  private JacksonIteratorDecoder(ObjectMapper mapper) {
+    this.mapper = mapper;
+  }
+
+  @Override
+  public Object decode(Response response, Type type) throws IOException {
+    if (response.status() == 404) return Util.emptyValueOf(type);
+    if (response.body() == null) return null;
+    Reader reader = response.body().asReader();
+    if (!reader.markSupported()) {
+      reader = new BufferedReader(reader, 1);
+    }
+    try {
+      // Read the first byte to see if we have any data
+      reader.mark(1);
+      if (reader.read() == -1) {
+        return null; // Eagerly returning null avoids "No content to map due to end-of-input"
+      }
+      reader.reset();
+      return new JacksonIterator<Object>(actualIteratorTypeArgument(type), mapper, response, reader);
+    } catch (RuntimeJsonMappingException e) {
+      if (e.getCause() != null && e.getCause() instanceof IOException) {
+        throw IOException.class.cast(e.getCause());
+      }
+      throw e;
+    }
+  }
+
+  private Type actualIteratorTypeArgument(Type type) {
+    if (!(type instanceof ParameterizedType)) {
+      throw new IllegalArgumentException("Not supported type " + type.toString());
+    }
+    ParameterizedType parameterizedType = (ParameterizedType) type;
+    if (!Iterator.class.equals(parameterizedType.getRawType())) {
+      throw new IllegalArgumentException("Not an iterator type " + parameterizedType.getRawType().toString());
+    }
+    return ((ParameterizedType) type).getActualTypeArguments()[0];
+  }
+
+  public static final class Factory {
+    public static JacksonIteratorDecoder create() {
+       return create(Collections.<Module>emptyList());
+    }
+    public static JacksonIteratorDecoder create(Iterable<Module> modules) {
+      return new JacksonIteratorDecoder(new ObjectMapper()
+          .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+          .registerModules(modules));
+    }
+    public static JacksonIteratorDecoder create(ObjectMapper objectMapper) {
+      return new JacksonIteratorDecoder(objectMapper);
+    }
+  }
+
+  private final class JacksonIterator<T> implements Iterator<T>, Closeable {
+    private final Response response;
+    private final JsonParser parser;
+    private final ObjectReader objectReader;
+
+    private T current;
+
+    private JacksonIterator(Type type, ObjectMapper mapper, Response response, Reader reader)
```

Comment:
one thing you'll notice in other feign code is that we infrequently use private access as the modifiers bloat bytecode and sharing packages isn't supported outside feign anyway. if you feel strongly about hiding things, making the type private but default modifiers on the ctor and symbols accomplishes the same.